packages/Kokkos/test/tstUserApplication.cpp

```diff
@@ -0,0 +1,1341 @@
+  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+//---------------------------------------------------------------------------//
+/*!
+ * \file   tstUserApplication.cpp
+ * \author Stuart Slattery
+ * \brief  UserApplication unit tests.
+ */
+//---------------------------------------------------------------------------//
+
+#include <DTK_ConfigDefs.hpp>
+#include <DTK_InputAllocators.hpp>
+#include <DTK_ParallelTraits.hpp>
+#include <DTK_UserApplication.hpp>
+#include <DTK_UserDataInterface.hpp>
+#include <DTK_UserFunctionRegistry.hpp>
+#include <DTK_View.hpp>
+
+#include <Kokkos_Core.hpp>
+
+#include <Teuchos_ScalarTraits.hpp>
+#include <Teuchos_UnitTestHarness.hpp>
+
+#include <memory>
+
+namespace UserAppTest
+{
+//---------------------------------------------------------------------------//
+// User class
+//---------------------------------------------------------------------------//
+template <class Scalar, class ExecutionSpace>
+class UserTestClass
+{
+  public:
+    UserTestClass( const int space_dim, const int size_1, const int size_2,
+                   const int offset, const Scalar init_val )
+        : _space_dim( space_dim )
+        , _size_1( size_1 )
+        , _size_2( size_2 )
+        , _offset( offset )
+        , _data( "test_class_data", size_1, space_dim )
+    { /* ... */
+    }
+
+  public:
+    int _space_dim;
+    size_t _size_1;
+    size_t _size_2;
+    int _offset;
+    Kokkos::View<Scalar **> _data;
+};
+
+//---------------------------------------------------------------------------//
+// User functions.
+//---------------------------------------------------------------------------//
+// Get the size parameters for building a node list.
+template <class Scalar, class ExecutionSpace>
+void nodeListSize( std::shared_ptr<void> user_data, int &space_dim,
+                   size_t &local_num_nodes, bool &has_ghosts )
+{
+    auto u = std::static_pointer_cast<UserTestClass<Scalar, ExecutionSpace>>(
+        user_data );
+
+    space_dim = u->_space_dim;
+    local_num_nodes = u->_size_1;
+    has_ghosts = true;
+}
+
+//---------------------------------------------------------------------------//
+// Get the data for a node list.
+template <class Scalar, class ExecutionSpace>
+void nodeListData(
+    std::shared_ptr<void> user_data,
+    DataTransferKit::View<DataTransferKit::Coordinate> coordinates,
+    DataTransferKit::View<bool> is_ghost_node )
+{
+    auto u = std::static_pointer_cast<UserTestClass<Scalar, ExecutionSpace>>(
+        user_data );
+
+    // The lambda does not properly capture class data so extract it.
+    int space_dim = u->_space_dim;
+    int size_1 = u->_size_1;
+    int offset = u->_offset;
```

Comment:
this was different - you can't pass the user class directly to device so we are extracting data here